Content.Shared/Weapons/Ranged/Systems/SharedGunSystem.Ballistic.cs

```diff
@@ -41,15 +41,50 @@ private void OnBallisticInteractUsing(EntityUid uid, BallisticAmmoProviderCompon
 
         if (GetBallisticShots(component) >= component.Capacity) return;
 
-        component.Entities.Add(args.Used);
-        component.Container.Insert(args.Used);
+        if (TryComp<BallisticAmmoProviderComponent>(args.Used, out var ammoComp))
+        {
+            if (GetBallisticShots(ammoComp) == 0) return;
+
+            while (GetBallisticShots(component) < component.Capacity && GetBallisticShots(ammoComp) > 0)
+            {
+                var xform = EntityManager.GetComponent<TransformComponent>(uid);
+
+                if (ammoComp.Entities.Count == 0) // If the entity doesn't have any spawned bullets
+                {
+                    if (GetBallisticShots(component) == 0) // If the gun doesn't have any ammo
```

Comment:
Same here, and also refer to my above comment on "unnecessarily iterating the container frequently".